pkg/aggregator/demultiplexer.go

```diff
@@ -0,0 +1,379 @@
+	"github.com/DataDog/datadog-agent/pkg/epforwarder"
+	"github.com/DataDog/datadog-agent/pkg/forwarder"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// DemultiplexerInstance is a shared global demultiplexer instance.
+// Initialized by InitAndStartAgentDemultiplexer or InitAndStartServerlessDemultiplexer,
+// could be nil otherwise. TODO(remy): remove this global instance in the future.
+//
+// Deprecated.
+var demultiplexerInstance Demultiplexer
+
+// Demultiplexer is composed of multiple samplers (check and time/dogstatsd)
+// a shared forwarder, the event platform forwarder, orchestrator data buffers
+// and other data that need to then be sent to the forwarders.
+// DemultiplexerOptions let you configure which forwarders have to be started.
+// They are not started automatically if `options.StartForwarders` is not
+// explicitly set to `true`.
+type Demultiplexer interface {
+	// General
+
+	Run()
+	Stop(flush bool)
+
+	// Aggregation API
+
+	// AddTimeSamples adds time samples processed by the DogStatsD server into a time sampler pipeline.
+	// The MetricSamples should have their hash computed.
+	// TODO(remy): not implemented yet.
+	AddTimeSamples(sample []metrics.MetricSample)
+	// AddCheckSample adds check sample sent by a check from one of the collectors into a check sampler pipeline.
+	// TODO(remy): not implemented yet.
+	AddCheckSample(sample metrics.MetricSample)
+	// FlushAggregatedData flushes all the aggregated data from the samplers to
+	// the serialization/forwarding parts.
+	FlushAggregatedData(start time.Time, waitForSerializer bool)
+	// Aggregator returns an aggregator that anyone can use. This method exists
+	// to keep compatibility with existing code while introducing the Demultiplexer,
+	// however, the plan is to remove it anytime soon.
+	//
+	// Deprecated.
+	Aggregator() *BufferedAggregator
+	// Serializer returns a serializer that anyone can use. This method exists
+	// to keep compatibility with existing code while introducing the Demultiplexer,
+	// however, the plan is to remove it anytime soon.
+	//
+	// Deprecated.
+	Serializer() serializer.MetricSerializer
+
+	// Senders API, mainly used by collectors/checks
+
+	GetSender(id check.ID) (Sender, error)
+	SetSender(sender Sender, id check.ID) error
+	DestroySender(id check.ID)
+	GetDefaultSender() (Sender, error)
+	ChangeAllSendersDefaultHostname(hostname string)
+	cleanSenders()
+}
+
+// AgentDemultiplexer is the demultiplexer implementation for the main Agent
+type AgentDemultiplexer struct {
+	sync.Mutex
+
+	// options are the options with which the demultiplexer has been created
+	options    DemultiplexerOptions
+	aggregator *BufferedAggregator
+	output
+	*senders
+}
+
+// DemultiplexerOptions are the options used to initialize a Demultiplexer.
+// Note that forwarders are not started automatically until `StartForwarders`
+// isn't set to `true`.
```

Comment:
"Aren't .. until .. isn't" is a bit confusing here.  I think, but I'm not sure, you mean
```suggestion
// Note that forwarders are not started automatically unless `StartForwarders`
// is set to `true`.
```

Reply:
Aha you're right 🙇  Rephrasing to: 
>  Note that forwarders are started automatically only if `StartForwarders` is set to `true`.